src/Emogrifier/CssConcatenator.php

```diff
@@ -0,0 +1,137 @@
+<?php
+
+namespace Pelago\Emogrifier;
+
+/**
+ * Facilitates building a CSS string by appending rule blocks one at a time, checking whether the media query,
+ * selectors, or declarations block are the same as those from the preceding block and combining blocks in such cases.
+ *
+ * @author Jake Hotson <jake.github@qzdesign.co.uk>
+ */
+class CssConcatenator
+{
+    /**
+     * CSS under construction.
+     *
+     * @var string
+     */
+    private $css = '';
+
+    /**
+     * Current media query string, e.g. "@media screen and (max-width:639px)" in the currently open media query block,
+     * or an empty string if not currently within a media query block.
+     *
+     * @var string
+     */
+    private $currentMedia = '';
+
+    /**
+     * Array whose keys are selectors for the rule block currently under construction (values are of no significance),
+     * or an empty array if no rule block under construction.
+     *
+     * @var int[]
+     */
+    private $currentSelectorsAsKeys = [];
+
+    /**
+     * Declarations for the rule block currently under construction,
+     * or an empty string if no rule block under construction.
+     *
+     * @var string
+     */
+    private $currentDeclarationsBlock = '';
+
+    /**
+     * Allow extending classes to call `parent::__construct()`.
+     */
+    public function __construct()
+    {
+    }
+
+    /**
+     * Append a declaration block to the CSS.
+     *
+     * @param string[]|string $selectors Array of selectors for the rule, e.g. ["ul", "ol", "p:first-child"],
+     *                                   or a single selector, e.g. "ul".
```

Comment:
Removed possibility of passing a string instead of an array for a single selector - it's just two extra characters to add `[...]` around it in the calling code and it was converted to an array anyway.